ethernet_init/ethernet_init.c

```diff
@@ -0,0 +1,382 @@
+    dm9051_config.int_gpio_num = spi_eth_module_config->int_gpio;
+    esp_eth_mac_t *mac = esp_eth_mac_new_dm9051(&dm9051_config, &mac_config);
+    esp_eth_phy_t *phy = esp_eth_phy_new_dm9051(&phy_config);
+#elif CONFIG_EXAMPLE_USE_W5500
+    eth_w5500_config_t w5500_config = ETH_W5500_DEFAULT_CONFIG(CONFIG_EXAMPLE_ETH_SPI_HOST, &spi_devcfg);
+    w5500_config.int_gpio_num = spi_eth_module_config->int_gpio;
+    esp_eth_mac_t *mac = esp_eth_mac_new_w5500(&w5500_config, &mac_config);
+    esp_eth_phy_t *phy = esp_eth_phy_new_w5500(&phy_config);
+#endif //CONFIG_EXAMPLE_USE_W5500
+    // Init Ethernet driver to default and install it
+    esp_eth_handle_t eth_handle = NULL;
+    esp_eth_config_t eth_config_spi = ETH_DEFAULT_CONFIG(mac, phy);
+    ESP_GOTO_ON_FALSE(esp_eth_driver_install(&eth_config_spi, &eth_handle) == ESP_OK, NULL, err, TAG, "SPI Ethernet driver install failed");
+
+    // The SPI Ethernet module might not have a burned factory MAC address, we can set it manually.
+    if (spi_eth_module_config->mac_addr != NULL) {
+        ESP_GOTO_ON_FALSE(esp_eth_ioctl(eth_handle, ETH_CMD_S_MAC_ADDR, spi_eth_module_config->mac_addr) == ESP_OK,
+                          NULL, err, TAG, "SPI Ethernet MAC address config failed");
+    }
+
+    if (mac_out != NULL) {
+        *mac_out = mac;
+    }
+    if (phy_out != NULL) {
+        *phy_out = phy;
+    }
+    return eth_handle;
+err:
+    if (eth_handle != NULL) {
+        esp_eth_driver_uninstall(eth_handle);
+    }
+    if (mac != NULL) {
+        mac->del(mac);
+    }
+    if (phy != NULL) {
+        phy->del(phy);
+    }
+    return ret;
+}
+#endif // CONFIG_EXAMPLE_USE_SPI_ETHERNET
+
+#if !defined(CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET)
+#define CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET 0
+#endif
+
+#if !defined(CONFIG_EXAMPLE_SPI_ETHERNETS_NUM)
+#define CONFIG_EXAMPLE_SPI_ETHERNETS_NUM 0
+#endif
+
+uint8_t eth_cnt_g = 0;
+esp_eth_handle_t *eth_handles_g = NULL;
+esp_eth_mac_t *mac_g[CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET + CONFIG_EXAMPLE_SPI_ETHERNETS_NUM];
+esp_eth_phy_t *phy_g[CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET + CONFIG_EXAMPLE_SPI_ETHERNETS_NUM];
+
+esp_err_t ethernet_init(esp_eth_handle_t *eth_handles_out[], uint8_t *eth_cnt_out)
+{
+    esp_err_t ret = ESP_OK;
+
+    ESP_ERROR_CHECK(esp_event_handler_register(ETH_EVENT, ESP_EVENT_ANY_ID, &eth_event_handler, NULL));
+
+#if CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET || CONFIG_EXAMPLE_USE_SPI_ETHERNET
+    ESP_GOTO_ON_FALSE(eth_handles_out != NULL && eth_cnt_out != NULL, ESP_ERR_INVALID_ARG,
+                      err, TAG, "invalid arguments: initialized handles array or number of interfaces");
+    eth_handles_g = calloc(SPI_ETHERNETS_NUM + INTERNAL_ETHERNETS_NUM, sizeof(esp_eth_handle_t));
+    ESP_GOTO_ON_FALSE(eth_handles_g != NULL, ESP_ERR_NO_MEM, err, TAG, "no memory");
+
+#if CONFIG_EXAMPLE_USE_INTERNAL_ETHERNET
+    eth_handles_g[eth_cnt_g] = eth_init_internal(&mac_g[eth_cnt_g], &phy_g[eth_cnt_g]);
```

Comment:
I'd suggest wrapping `mac` and `phy` pointers to a structure together with `eth_handles_g`. You can also use c-style sub/super struct with `container-of()` operator.

Reply:
Created ```struct eth_device``` to wrap Mac, phy and eth_handle.